tls/s2n_tls13_secrets.c

```diff
@@ -0,0 +1,502 @@
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+#include "tls/s2n_tls13_secrets.h"
+
+#include "tls/s2n_connection.h"
+#include "tls/s2n_key_log.h"
+#include "tls/s2n_tls13_handshake.h"
+#include "utils/s2n_bitmap.h"
+
+#define S2N_MAX_HASHLEN SHA384_DIGEST_LENGTH
+
+#define CONN_HMAC_ALG(conn) ((conn)->secure.cipher_suite->prf_alg)
+#define CONN_SECRETS(conn)  ((conn)->secrets.tls13)
+#define CONN_HASHES(conn)   ((conn)->handshake.hashes)
+
+#define CONN_SECRET(conn, secret) ( \
+    (struct s2n_blob) { .data = CONN_SECRETS(conn).secret, .size = s2n_get_hash_len(CONN_HMAC_ALG(conn))} )
+#define CONN_HASH(conn, hash) ( \
+    (struct s2n_blob) { .data = CONN_HASHES(conn)->hash, .size = s2n_get_hash_len(CONN_HMAC_ALG(conn))} )
+#define CONN_FINISHED(conn, mode) ( \
+    (struct s2n_blob) { .data = (conn)->handshake.mode##_finished, .size = s2n_get_hash_len(CONN_HMAC_ALG(conn))})
+
+/**
+ *= https://tools.ietf.org/rfc/rfc8446#section-7.1
+ *# If a given secret is not available, then the 0-value consisting of a
+ *# string of Hash.length bytes set to zeros is used.
+ */
+static uint8_t zero_value_bytes[S2N_MAX_HASHLEN] = { 0 };
+#define ZERO_VALUE(hmac_alg) ( \
+    (const struct s2n_blob) { .data = zero_value_bytes, .size = s2n_get_hash_len(hmac_alg)})
+
+/**
+ *= https://tools.ietf.org/rfc/rfc8446#section-7.1
+ *# Note that in some cases a zero-
+ *# length Context (indicated by "") is passed to HKDF-Expand-Label
+ */
+#define EMPTY_CONTEXT(hmac_alg) ( \
+    (const struct s2n_blob) { .data = s2n_get_empty_context(hmac_alg), .size = s2n_get_hash_len(hmac_alg)})
+
+static uint8_t s2n_get_hash_len(s2n_hmac_algorithm hmac_alg)
+{
+    uint8_t hash_size = 0;
+    if (s2n_hmac_digest_size(hmac_alg, &hash_size) != S2N_SUCCESS) {
+        return 0;
+    }
+    return hash_size;
+}
+
+static uint8_t *s2n_get_empty_context(s2n_hmac_algorithm hmac_alg)
+{
+    static uint8_t sha256_empty_digest[S2N_MAX_HASHLEN] = { 0 };
+    static uint8_t sha384_empty_digest[S2N_MAX_HASHLEN] = { 0 };
+
+    switch(hmac_alg) {
+        case S2N_HMAC_SHA256:
+            return sha256_empty_digest;
+        case S2N_HMAC_SHA384:
+            return sha384_empty_digest;
+        default:
+            return NULL;
+    }
+}
+
+static s2n_hmac_algorithm supported_hmacs[] = {
+        S2N_HMAC_SHA256,
+        S2N_HMAC_SHA384
+};
+
+S2N_RESULT s2n_tls13_secrets_init()
+{
+    DEFER_CLEANUP(struct s2n_hash_state hash = { 0 }, s2n_hash_free);
+    RESULT_GUARD_POSIX(s2n_hash_new(&hash));
+
+    s2n_hash_algorithm hash_alg = S2N_HASH_NONE;
+    for (size_t i = 0; i < s2n_array_len(supported_hmacs); i++) {
+        s2n_hmac_algorithm hmac_alg = supported_hmacs[i];
+        struct s2n_blob digest = EMPTY_CONTEXT(hmac_alg);
+
+        RESULT_GUARD_POSIX(s2n_hmac_hash_alg(hmac_alg, &hash_alg));
```

Comment:
I tried to do this, but we'd still have to convert hash->hmac to retrieve the right empty context. Every other time we need an empty context we only have the hmac, so the method to retrieve an empty context takes an hmac instead of a hash.

It's frustrating that those two enums are essentially interchangeable, but not actually interchangeable :(